python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -861,10 +869,18 @@ def _extractIngestInfo(self, path: Union[str, ButlerURI], ref: DatasetRef, *,
                 # Work out the path in the datastore from an absolute URI
                 # This is required to be within the datastore.
                 pathInStore = srcUri.relative_to(self.root)
-                if pathInStore is None:
+                if pathInStore is None and transfer is None:
                     raise RuntimeError(f"Unexpectedly learned that {srcUri} is "
                                        f"not within datastore {self.root}")
-                tgtLocation = self.locationFactory.fromPath(pathInStore)
+                if pathInStore:
+                    tgtLocation = self.locationFactory.fromPath(pathInStore)
+                elif transfer == "split":
+                    # Outside the datastore but treat that as a direct ingest
+                    # instead.
+                    tgtLocation = None
+                else:
+                    raise RuntimeError(f"Unexpected transfer mode encountered: {transfer} for"
+                                       f" URI {srcUri}")
```

Comment:
it seems like this whole if block is combining two different bits of logic together. It might be the right thing, but I am having a bit of trouble following it, not knowing what some of the results of calls will be.

You have changed to `pathInStore is None and transfer is None` is this implying that you are using something like True, False, and None as conditions in normal operation? Is it ok that if `pathInStore` is None, transfer is not None, for the `if pathInStore` to evaluate to false based on None being false-y?

I think I understand what you are doing and I think it is fine, but I am sufficiently uncertain, that I want to highlight it for a second look by you.

Reply:
@TallJimbo used `transfer=None` to mean "the files are already in the right place" and no transfer is needed. For `split` mode I need to handle the case where some are in and some are out. For the files inside the target location is relative to the datastore root. For the files outside there is also no transfer but that is "direct" mode and is handled by setting the target location to None (so the code below knows to use the source URI directly). It is combining two bits of logic but the `None` mode is most of the logic and it didn't seem like it was reasonable to add a new `split` path as a separate check that would duplicate the `None` -- the only change is to set that target URI to `None` in split mode.